### PR TITLE
:sparkles: Add support for regexp in profile activation kubeContext

### PR DESCRIPTION
**Pull request by charlyx**
Hi 👋

This PR adds support for regexp in profile activation via kubeContext as specified in #1677 
I tried to avoid breaking any existing behaviour.

Thanks for your review 🙏 

PS: thanks @pyaillet for his help 😄

### CODE REVIEW

File: pkg/skaffold/schema/profiles.go [hunk 1]
```diff
@@ -124,7 +125,13 @@ func satisfies(expected, actual string) bool {
 	if strings.HasPrefix(expected, "!") {
 		return actual != expected[1:]
 	}
-	return actual == expected
+	matcher, err := re.Compile(expected)
+	if err != nil {
+		logrus.Infof("Not a regexp: %s, falling back to string equals", expected)
```
Comment (balopat) [inline pkg/skaffold/schema/profiles.go]:
```suggestion
		logrus.Infof("profile activation criteria '%s' is not a valid regexp, falling back to string equals", expected)
```

File: pkg/skaffold/schema/profiles.go [hunk 2]
```diff
@@ -124,7 +125,13 @@ func satisfies(expected, actual string) bool {
 	if strings.HasPrefix(expected, "!") {
 		return actual != expected[1:]
```
Comment (balopat) [inline pkg/skaffold/schema/profiles.go]: I think we should support the `!` prefix for regular expressions too, e.g.:

`!dev-.*` should activate on all that don't match it

File: pkg/skaffold/schema/profiles.go [hunk 3]
```diff
@@ -122,9 +123,23 @@ func isKubeContext(kubeContext string) (bool, error) {
...
+	}
+
+	return matches(expected, actual)
+}
+
+func matches(expected, actual string) bool {
+	matcher, err := re.Compile(expected)
+
+	if err != nil {
+		logrus.Infof("profile activation criteria '%s' is not a valid regexp, falling back to string", expected)
```
Comment (balopat) [inline pkg/skaffold/schema/profiles.go]:
```suggestion
		logrus.Infof("profile activation criteria '%s' is not a valid regexp, falling back to string: %v", expected, err)
```

File: pkg/skaffold/schema/profiles_test.go [hunk 4]
```diff
@@ -377,8 +377,10 @@ func TestActivatedProfiles(t *testing.T) {
 				{Name: "activated", Activation: []latest.Activation{{KubeContext: "prod-context"}}},
 				{Name: "not-activated", Activation: []latest.Activation{{KubeContext: "dev-context"}}},
 				{Name: "also-activated", Activation: []latest.Activation{{KubeContext: "!dev-context"}}},
+				{Name: "activated-regexp", Activation: []latest.Activation{{KubeContext: "prod-.*"}}},
+				{Name: "not-activated-regexp", Activation: []latest.Activation{{KubeContext: "dev-.*"}}},
```
Comment (balopat) [inline pkg/skaffold/schema/profiles_test.go]: we should cover the !(regex) and invalid regex cases too.

Comment (charlyx) [reply pkg/skaffold/schema/profiles_test.go]: Should I create another profile so I can test those cases more easily?

For instance, in order to test the invalid regex I would need a profile that matches when comparing strings, I guess. Also, is there any character limitation for the profile? I couldn't find any in [create_context.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/config/create_context.go)

Regarding the `!regexp` I'm not sure how to test it... Shall we assume that if the regexp is valid we're testing the regexp matching and not the string comparison so I can make a simple test?

**Comment by dgageot** [comment]
Hi @charlyx! Do you need help to finish this PR?

**Comment by charlyx** [comment]
Hi @dgageot 👋 

Thanks for asking!
Actually, I asked a question a few weeks ago: https://github.com/GoogleContainerTools/skaffold/pull/2065#discussion_r284616690

File: pkg/skaffold/schema/profiles_test.go (hunk 4 above)
Comment (dgageot) [reply pkg/skaffold/schema/profiles_test.go]: Feel free to add as many profiles as needed. Also, simple tests are better!

**Comment by dgageot** [comment]
Hi @charlyx, I'd like to see this PR merge. Anything I can do?

**Comment by dgageot** [comment]
@charlyx I'm going to merge this PR and add the tests

**Comment by charlyx** [comment]
Sorry I couldn’t make it on time David. 
I’ve been quite busy lately. 

Thanks for the merge!

**Pull request merged**